src/main/java/org/opensearch/plugin/insights/core/reader/LocalIndexReader.java

```diff
@@ -0,0 +1,130 @@
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormatter;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.Client;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.index.IndexNotFoundException;
+import org.opensearch.index.query.MatchQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.index.query.RangeQueryBuilder;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+/**
+ * Local index reader for reading query insights data from local OpenSearch indices.
+ */
+public final class LocalIndexReader implements QueryInsightsReader {
+    /**
+     * Logger of the local index reader
+     */
+    private final Logger logger = LogManager.getLogger();
+    private final Client client;
+    private DateTimeFormatter indexPattern;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    /**
+     * Constructor of LocalIndexReader
+     *
+     * @param client OS client
+     * @param indexPattern the pattern of index to read from
+     * @param namedXContentRegistry for parsing purposes
+     */
+    public LocalIndexReader(final Client client, final DateTimeFormatter indexPattern, final NamedXContentRegistry namedXContentRegistry) {
+        this.indexPattern = indexPattern;
+        this.client = client;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    /**
+     * Getter of indexPattern
+     *
+     * @return indexPattern
+     */
+    public DateTimeFormatter getIndexPattern() {
+        return indexPattern;
+    }
+
+    /**
+     * Setter of indexPattern
+     *
+     * @param indexPattern index pattern
+     * @return the current LocalIndexReader
+     */
+    public LocalIndexReader setIndexPattern(DateTimeFormatter indexPattern) {
+        this.indexPattern = indexPattern;
+        return this;
+    }
+
+    /**
+     * Export a list of SearchQueryRecord from local index
+     *
+     * @param from start timestamp
+     * @param to end timestamp
+     * @return list of SearchQueryRecords whose timestamps fall between from and to
+     */
+    @Override
+    public List<SearchQueryRecord> read(final String from, final String to) {
+        List<SearchQueryRecord> records = new ArrayList<>();
+        if (from == null || to == null) {
+            return records;
+        }
+        final DateTime start = DateTime.parse(from);
+        DateTime end = DateTime.parse(to);
+        if (end.compareTo(DateTime.now(DateTimeZone.UTC)) > 0) {
+            end = DateTime.now(DateTimeZone.UTC);
+        }
+        DateTime curr = start;
+        while (curr.compareTo(end.plusDays(1).withTimeAtStartOfDay()) < 0) {
```

Comment:
Note: We are making assumptions on the index pattern that it will rollover every day. We should add this assumption with a check in the configuration API as well.